kale/pipeline/mida_trainer.py

```diff
@@ -0,0 +1,201 @@
+# =============================================================================
+# Author: Shuo Zhou, shuo.zhou@sheffield.ac.uk
+# =============================================================================
+
+"""Implementation of Transformer -> Maximum independence domain adaptation -> Estimator Pipeline
+
+References:
+    [1] Yan, K., Kou, L. and Zhang, D., 2018. Learning domain-invariant subspace using domain features and
+        independence maximization. IEEE transactions on cybernetics, 48(1), pp.288-299.
+"""
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.model_selection import GridSearchCV, LeaveOneGroupOut, ParameterGrid
+from sklearn.preprocessing import FunctionTransformer
+
+from ..embed.factorization import MIDA
+
+
+class MIDATrainer(BaseEstimator, ClassifierMixin):
+    """Trainer of pipeline: Transformer (optional) -> Maximum independence domain adaptation -> Estimator
+
+    Args:
+        estimator (BaseEstimator): Estimator object implementing 'fit'
+        estimator_param_grid (dict, optional): Grids for searching the optimal hyperparameters of the estimator.
+            Defaults to None.
+        transformer (BaseEstimator, optional): Transformer object implementing 'fit_transform'. Defaults to None.
+        mida_param_grid (dict, optional): Grids for searching the optimal hyperparameters of MIDA. Defaults to None.
+        transformer_param_grid (dict, optional): Grids for searching the optimal hyperparameters of the transformer.
+            Defaults to None.
+    """
+
+    def __init__(
+        self,
+        estimator,
+        estimator_param_grid=None,
+        transformer=None,
+        mida_param_grid=None,
+        transformer_param_grid=None,
+    ):
+        self.estimator = estimator
+        self.transformer = transformer
+        self.mida_param_grid = mida_param_grid
+        self.transformer_param_grid = transformer_param_grid
+        self.estimator_param_grid = estimator_param_grid
+        self.best_transformer = None
```

Comment:
initialize the `best` attributes during `fit` rather than `__init__`. and also use standard way on how to define attributes during fit such as `best_mida_`